src/qr.jl

```diff
@@ -2,20 +2,203 @@ _thin_must_hold(thin) =
     thin || throw(ArgumentError("For the sake of type stability, `thin = true` must hold."))
 import Base.qr
 
-function qr(A::StaticMatrix, pivot::Type{Val{true}}; thin::Bool=true)
+
+"""
+    qr(A::StaticMatrix, pivot=Val{false}; thin=true) -> Q, R, [p]
+
+Compute the QR factorization of `A` such that `A = Q*R` or `A[:,p] = Q*R`, see [`qr`](@ref).
+This function is not support `thin=false` keyword option due to type inference instability.
+To use this option call `StaticArrays._qr(Size(A), A, pivot, Val{false})` instead.
+"""
+@inline function qr(A::StaticMatrix, pivot::Union{Type{Val{false}}, Type{Val{true}}} = Val{false}; thin::Bool=true)
     _thin_must_hold(thin)
-    Q0, R0, p0 = Base.qr(Matrix(A), pivot)
-    T = arithmetic_closure(eltype(A))
-    QT = similar_type(A, T, Size(diagsize(A), diagsize(A)))
-    RT = similar_type(A, T)
-    PT = similar_type(A, Int, Size(Size(A)[2]))
-    QT(Q0), RT(R0), PT(p0)
+    return _qr(Size(A), A, pivot, Val{true})
 end
-function qr(A::StaticMatrix, pivot::Type{Val{false}}; thin::Bool=true)
-    _thin_must_hold(thin)
-    Q0, R0 = Base.qr(Matrix(A), pivot)
-    T = arithmetic_closure(eltype(A))
-    QT = similar_type(A, T, Size(diagsize(A), diagsize(A)))
-    RT = similar_type(A, T)
-    QT(Q0), RT(R0)
+
+
+@generated function _qr(::Size{sA}, A::StaticMatrix{<:Any, <:Any, TA}, pivot::Union{Type{Val{false}}, Type{Val{true}}} = Val{false}, thin::Union{Type{Val{false}}, Type{Val{true}}} = Val{true}) where {sA, TA}
+
+    isthin = thin <: Type{Val{true}}
```

Comment:
This can be expressed a little more straightforwardly as `thin isa Val{true}` (same elsewhere)

Reply:
`isa` in `@generated` functions is not worked.

I had try `===` which should work, but test gives an error, because in [line](https://github.com/denius/StaticArrays.jl/blob/qr-wopivoting/src/qr.jl#L29) `if` condition can't rightly recognize `pivot`.

Thus there is remains only `<:` and `==` correct variants. And I prefer `<:` because it deal with types.

Reply:
Ah right of course - this is a generated function body so my suggestion to use `isa` is wrong.

I'd say it's quite acceptable and appropriate use `==` here because types are values in julia. In this case `<:` also works because `Val{true}` is a concrete type (ie has no subtypes), but as the subtype operator it suggests different (type partial ordering) semantics.

Reply:
Ok, I get it, thanks

Reply:
Thanks so much for changing all these small things!

I just realized I was a little extra confused here because in julia-0.7-dev the `qr` function has been changed to take a `Val{true}()` instance rather than a `Val{true}` type.  But let's worry about that another time.

Reply:
Yeah, that's my fault...

Reply:
Missed one eh? :-)